### PR TITLE
pythonPackages.tensorflow: 1.3.1 -> 1.4.1

### PR DESCRIPTION
**Pull request by jyp**
Also revert to a wheel-based build (the bazel-based build is broken
and it is unclear how to repair it)

###### Motivation for this change

Two motivations:

- Tensorflow does not currently build in master
- Upstream update

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by jyp** [comment]
@FRidh Would you think that this version is acceptable? (@abbradar still has not given any sign of activity)

**Comment by lukeadams** [comment]
@jyp [Apparently Arch is able to build 1.5x from git.](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=tensorflow-git) Might try to dig into it this weekend – when I try to build 1.5, bazel throws errors about jemalloc caused by various scoping issues which I'd imagine are straightforward to fix.
[My 1.5x branch](https://github.com/lukeadams/nixpkgs/tree/tf-update)
[Build log](https://gist.github.com/lukeadams/f7f1ae6481c9abd081fdb346791fb4c2)

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 1]
```diff
@@ -1,187 +1,112 @@
...
-    mkdir -p $out/${python.sitePackages}
-    bazel --output_base="$(pwd)/output" build $bazelFlags $bazelTarget
-
-    bazel-bin/tensorflow/tools/pip_package/build_pip_package $PWD/dist
-  '';
-
-  # tensorflow depends on tensorflow_tensorboard, which cannot be
+  # cudatoolkit is split (see https://github.com/NixOS/nixpkgs/commit/bb1c9b027d343f2ce263496582d6b56af8af92e6)
+  # However this means that libcusolver is not loadable by tensor flow. So we undo the split here.
+  cudatoolkit_joined = symlinkJoin {
```
Comment (proger) [inline pkgs/development/python-modules/tensorflow/default.nix]: This needs to be wrapped into `if cudaSupport` or smth like that

File: pkgs/development/python-modules/tensorflow/default.nix [hunk 2]
```diff
@@ -1,187 +1,112 @@
...
+              cudatoolkit.lib ];};
+
+  src = let dls = import ./tf1.4.1-hashes.nix;
+    in
+    fetchurl (
+      if stdenv.isDarwin then
+        if isPy3k then
+          dls.mac_py_2_cpu
+        else
+          dls.mac_py_3_cpu
```
Comment (proger) [inline pkgs/development/python-modules/tensorflow/default.nix]: Seems to be a mixup here.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -21735,11 +21735,10 @@ EOF
   tensorflow-tensorboard = callPackage ../development/python-modules/tensorflow-tensorboard { };
 
   tensorflow = callPackage ../development/python-modules/tensorflow rec {
-    bazel = pkgs.bazel_0_4;
     cudaSupport = pkgs.config.cudaSupport or false;
     inherit (pkgs.linuxPackages) nvidia_x11;
-    cudatoolkit = pkgs.cudatoolkit8;
-    cudnn = pkgs.cudnn6_cudatoolkit8;
+    cudatoolkit = pkgs.cudatoolkit9;
+    cudnn = pkgs.cudnn_cudatoolkit9;
```
Comment (proger) [inline pkgs/top-level/python-packages.nix]: I had to remove these entirely so my darwin build goes through.

**Review by proger — COMMENTED** [review]
Thanks for your work!

I did some minor tweaks to make your PR work on Darwin, here's the hello world output:
```
[nix-shell:~]$ ipython --quick
Python 3.6.4 (default, Jan 17 2018, 00:49:01) 
Type 'copyright', 'credits' or 'license' for more information
IPython 6.2.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import tensorflow as tf
/nix/store/wc4m3g4grivhy26sdck44jxwkvz1j17n-python3-3.6.4/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: compiletime version 3.5 of module 'tensorflow.python.framework.fast_tensor_util' does not match runtime version 3.6
  return f(*args, **kwds)

In [2]: hello = tf.constant('Hello, TensorFlow!')

In [3]: sess = tf.Session()
2018-02-09 18:21:45.788566: I tensorflow/core/platform/cpu_feature_guard.cc:137] Your CPU supports instructions that this TensorFlow binary was not compiled to use: SSE4.1 SSE4.2 AVX AVX2 FMA

In [4]: print(sess.run(hello))
b'Hello, TensorFlow!'

```

The [word2vec_basic](https://github.com/tensorflow/tensorflow/blob/r1.4/tensorflow/examples/tutorials/word2vec/word2vec_basic.py) example seems to work well too.

Tensorboard seems to be broken though:

```
[nix-shell:~]$ tensorboard
Traceback (most recent call last):
  File "/nix/store/jvckx7illg4s3y64443swlgbsmv0n1w6-python3.6-tensorflow-1.4.1/bin/.tensorboard-wrapped", line 8, in <module>
    from tensorboard.main import main
ModuleNotFoundError: No module named 'tensorboard'
```

**Comment by lukeadams** [comment]
@jyp (continuation from #31492)
> I have another PR with 1.4 which should be mergeable now. If you support the revert to a wheel-based build please comment about it in the PR.

I'm not against getting this merged in the short term, but ideally we would trace down why the bazel build fails (source build would make #31046 straightforward to resolve). Easy to undo this if/when we get build from source working.

This method is nice since Bazel on Darwin is broken.

File: pkgs/development/python-modules/tensorflow/default.nix (hunk 1 above)
Comment (jyp) [reply pkgs/development/python-modules/tensorflow/default.nix]: Is that really so? I thought that `nix` had lazy evaluation.

**Comment by jyp** [comment]
@lukeadams "Ideally" yes.  If someone can fix it I'll be happy to see the bazel build restored.  In practice tensorflow has been broken for three months in master now, hence this PR.

**Comment by jyp** [comment]
@proger Tensorboard is incompatible with python 3.6. https://github.com/tensorflow/tensorboard/issues/427

File: pkgs/top-level/python-packages.nix (hunk 3 above)
Comment (jyp) [reply pkgs/top-level/python-packages.nix]: This comes from master --- I'd suggest to fix this in another patch.

**Comment by jyp** [comment]
@proger Thanks a lot for your review. I acted on your suggestions but I've left the top-level definition as it was to minimize the diff with master. We can change that one separately. (It's hard enough to get the simplest PR merged.)

**Comment by lukeadams** [comment]
[This patch is all that's necessary to build on Darwin:](https://gist.github.com/lukeadams/920698566328d7fd8eeda7116f87b9a7)
```patch
From 5cbbb34a43cc431aeb40591042f3d61c31c88085 Mon Sep 17 00:00:00 2001
From: Luke Adams <luke.adams@belljar.io>
Date: Sat, 10 Feb 2018 10:18:50 -0600
Subject: [PATCH] tensorflow: move cudatoolkit_joined to top-level for clarity

allows for simplification of postFixup
---
 .../python-modules/tensorflow/default.nix          | 24 +++++++++++-----------
 1 file changed, 12 insertions(+), 12 deletions(-)

diff --git a/pkgs/development/python-modules/tensorflow/default.nix b/pkgs/development/python-modules/tensorflow/default.nix
index 5c5d98e7ace..ff34d02c4fa 100644
--- a/pkgs/development/python-modules/tensorflow/default.nix
+++ b/pkgs/development/python-modules/tensorflow/default.nix
@@ -31,8 +31,18 @@ assert ! (stdenv.isDarwin && cudaSupport);
 # project's build system is an arcane beast based on
 # bazel. Untangling it and building the wheel from source is an open
 # problem.
+let
+  # cudatoolkit is split (see https://github.com/NixOS/nixpkgs/commit/bb1c9b027d343f2ce263496582d6b56af8af92e6)
+  # However this means that libcusolver is not loadable by tensor flow. So we undo the split here.
+  cudatoolkit_joined = symlinkJoin {
+    name = "unsplit_cudatoolkit";
+    paths = [
+      cudatoolkit.out
+      cudatoolkit.lib
+    ];
+   };
 
-buildPythonPackage rec {
+in buildPythonPackage rec {
   pname = "tensorflow";
   version = "1.4.1";
   name = "${pname}-${version}";
@@ -82,17 +92,7 @@ buildPythonPackage rec {
   # patchelf --shrink-rpath will remove the cuda libraries.
   postFixup = let
     rpath = stdenv.lib.makeLibraryPath
-      (if cudaSupport then
-        # cudatoolkit is split (see https://github.com/NixOS/nixpkgs/commit/bb1c9b027d343f2ce263496582d6b56af8af92e6)
-        # However this means that libcusolver is not loadable by tensor flow. So we undo the split here.
-        (let cudatoolkit_joined = symlinkJoin {
-          name = "unsplit_cudatoolkit";
-          paths = [ cudatoolkit.out
-                    cudatoolkit.lib ];};
-         in [ stdenv.cc.cc.lib zlib cudatoolkit_joined cudnn nvidia_x11 ])
-      else
-        [ stdenv.cc.cc.lib zlib ]
-      );
+      ([ stdenv.cc.cc.lib zlib ] ++ lib.optionals cudaSupport [ cudatoolkit_joined cudnn nvidia_x11 ]);
   in
   ''
     rrPath="$out/${python.sitePackages}/tensorflow/:${rpath}"
-- 
2.15.1
```

File: pkgs/development/python-modules/tensorflow/default.nix (hunk 1 above)
Comment (proger) [reply pkgs/development/python-modules/tensorflow/default.nix]: I think all derivation (or `stdenv.mkDerivation`?) attributes are forced to become environment variables in a generated build script.

**Comment by jyp** [comment]
I have applied the patch of @lukeadams 

Could this PR be considered for merging? @FRidh ?

**Comment by jyp** [comment]
Tensorflow is still not building ( #31492 ), even after merging the PR of @abbradar .  Can we please merge this PR until the bazel build works? @FRidh ?

**Comment by abbradar** [comment]
That's very strange as I'm using it right now. Can you show your error?

**Comment by abbradar** [comment]
Ah, I see failure on Hydra (it was down today morning so I didn't check it first). The error is very strange; for now my theory is that something in nixpkgs has updated since the last successful build (https://hydra.nixos.org/build/69832610). Currently tensorflow doesn't build because of unrelated reasons on master; I'll see if I can reproduce the issue.

**Comment by jyp** [comment]
Thanks! Please report progress on #31492, I'll ping you from there.

**Comment by lukeadams** [comment]
Just wanted to mention: building from source prevents the use of TF on Darwin until #30590 / #32080 are fixed.

**Comment by abbradar** [comment]
@lukeadams Yeah, I want to tackle that next -- or return wheel-based build specifically for Darwin.

**Comment by abbradar** [comment]
I've tried to set up a Mac OS X VirtualBox machine but apparently AMD processors are not supported, and on an image from https://forum.amd-osx.com/ mouse and keyboard do not work.

@LnL7 Can you perhaps set a Darwin Hydra job for my `tensorflow` branch again? For now I've reset it to be just `master`.

**Comment by FRidh** [comment]
@abbradar in case it is not possible to get the source builds working entirely on all supported archs, I think we should still provide the wheel-based solution, at least for 18.03, for all archs.

In that case, @jyp, I recommend including the script you used as an actual script in the tree. Also, instead of bothering to write .nix you can just output .json.

**Comment by abbradar** [comment]
@FRidh Yeah, that's my opinion too -- I thought I'd try to tackle this one on holidays before giving up completely.

**Comment by abbradar** [comment]
@jyp I feel too defeated by Darwin for now D: Can you rebase this pull request and make it be used only for Darwin? I can do it by myself if you don't have enough time. Thanks in advance!

**Comment by jyp** [comment]
@abbradar I am completely swamped at the moment so it's great if you can take this (I am not even a darwin user btw). Additionally I wouldn't even know how to apply the advice of @FRidh.

**Pull request closed without merging**